src/container/FluxContainer.js

```diff
@@ -98,116 +86,43 @@ const DEFAULT_OPTIONS = {
-
-    constructor(props: Props, context: any) {
-      super(props, context);
-      this._fluxContainerSubscriptions = new FluxContainerSubscriptions();
-      this._fluxContainerSubscriptions.setStores(getStores(props));
-      this._fluxContainerSubscriptions.addListener(() => {
-        this.setState(
-          (prevState, currentProps) => getState(
-            prevState,
-            currentProps,
-            context,
-          ),
-        );
-      });
-      const calculatedState = getState(undefined, props, context);
-      this.state = {
-        ...(this.state || {}),
-        ...calculatedState,
-      };
-    }
-
-    componentWillReceiveProps(nextProps: any, nextContext: any): void {
-      if (super.componentWillReceiveProps) {
-        super.componentWillReceiveProps(nextProps, nextContext);
-      }
-
-      if (realOptions.withProps || realOptions.withContext) {
-        // Update both stores and state.
-        this._fluxContainerSubscriptions.setStores(
-          getStores(nextProps, nextContext),
-        );
-        this.setState(prevState => getState(prevState, nextProps, nextContext));
-      }
-    }
-
-    componentWillUnmount() {
-      if (super.componentWillUnmount) {
-        super.componentWillUnmount();
-      }
-
-      this._fluxContainerSubscriptions.reset();
-    }
+class FluxContainer extends Component<DefaultProps, Props, State> {
+  state: State;
+  constructor(props: Props, context: any) {
+    super(props, context);
+    this._fluxContainerSubscriptions = new FluxContainerSubscriptions();
+    this._fluxContainerSubscriptions.setStores(this.getStores());
+    this._fluxContainerSubscriptions.addListener(() => {
+      this.setState((prevState) => this.calculateState(prevState));
+    });
+    this.state = this.calculateState();
   }
 
-  // Make sure we override shouldComponentUpdate only if the pure option is
-  // specified. We can't override this above because we don't want to override
-  // the default behavior on accident. Super works weird with react ES6 classes.
-  const container = realOptions.pure
-    ? createPureComponent(ContainerClass)
-    : ContainerClass;
+  getStores(): Array<FluxStore> {
+    return abstractMethod('FluxContainer', 'getStores');
+  }
 
-  // Update the name of the container before returning
-  const componentName = Base.displayName || Base.name;
-  container.displayName = 'FluxContainer(' + componentName + ')';
-  return container;
-}
+  calculateState(): State {
+    return abstractMethod('FluxContainer', 'calculateState');
+  }
 
-function createPureComponent<DefaultProps, Props, State>(
-  BaseComponent: ReactClass<Props>
-): ReactClass<Props> {
-  class PureComponent extends BaseComponent {
-    shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
-      return (
-        !shallowEqual(this.props, nextProps) ||
-        !shallowEqual(this.state, nextState)
-      );
-    }
+  componentWillReceiveProps(nextProps: Props, nextContext: any): void {
```

Comment:
This method is called frequently, it happens before `shouldComponentUpdate()` so even if we are dealing with pure components that can return false from `shouldComponentUpdate()` in most situations `componentWillReceiveProps()` should still be considered a hot path.

Placing `calculateState()` in here on every container hurts performance a ton. It's strongly discouraged from using props in containers so this becomes a massive time sink for the majority of containers which do not need props (although there are some valid use cases which is why this `withProps` escape hatch exists.)

The reason for `calculateState()` being static was so that people could not access `this.props` and accidentally use props without making sure the container correctly hooks into `componentWillReceiveProps()`.